images/hook/entrypoint.sh

```diff
@@ -1,44 +1,112 @@
+  VERSION=$(kubectl exec -it "${MAYA_POD}" mayactl version -nopenebs | grep ^Version | cut -d" " -f2 | perl -pe's/(\d+.\d+.\d+).*/$1/')
 
-    echo "--> Checking: $RIG_CHANGESET"
-    if rig status ${RIG_CHANGESET} --retry-attempts=1 --retry-period=1s; then
-        exit 0
-    fi
+  echo "$VERSION"
+}
 
-    echo "--> Starting upgrade, changeset: $RIG_CHANGESET"
-    rig cs delete --force -c cs/${RIG_CHANGESET}
+# Verifies that the control plane components are of the expected version
+function check_control_plane() {
+  echo "Checking control plane for version=$1"
 
-    # TODO(r0mant): As this is the first release of OpenEBS as a Gravity application,
-    #               the current "upgrade" procedure assumes that OpenEBS is not
-    #               yet installed and thus just creates all OpenEBS resources
-    #               from scratch.
-    #               In the future, we'll need to develop a proper upgrade procedure
-    #               for OpenEBS components and its pools/volumes like described in
-    #               https://github.com/openebs/openebs/tree/master/k8s/upgrades.
+  CONTROL_PLANE_COMPONENTS=control_plane_components.txt
+  kubectl get pods -n openebs -l openebs.io/version="$1" > $CONTROL_PLANE_COMPONENTS
 
-    echo "--> Creating OpenEBS resources"
-    rig upsert -f /var/lib/gravity/resources/openebs-operator.yaml --debug
+  echo "Found control plane components:"
+  cat $CONTROL_PLANE_COMPONENTS
 
-    echo "--> Checking status"
-    rig status ${RIG_CHANGESET} --retry-attempts=120 --retry-period=1s --debug
+  grep 'provisioner.*Running' $CONTROL_PLANE_COMPONENTS >/dev/null &&
+    grep 'admission-server.*Running' $CONTROL_PLANE_COMPONENTS >/dev/null &&
+    grep 'maya-apiserver.*Running' $CONTROL_PLANE_COMPONENTS >/dev/null &&
+    grep 'ndm.*Running' $CONTROL_PLANE_COMPONENTS >/dev/null &&
+    grep 'snapshot-operator.*Running' $CONTROL_PLANE_COMPONENTS >/dev/null
 
-    echo "--> Freezing"
-    rig freeze
+  return $?
+}
 
-elif [ $1 = "rollback" ]; then
+if [ "$1" = "update" ]; then
+  TO_VERSION=2.2.0
 
-    echo "--> Reverting changeset $RIG_CHANGESET"
-    rig revert
-    rig cs delete --force -c cs/${RIG_CHANGESET}
+  echo "--> Checking: $RIG_CHANGESET"
+  if rig status "${RIG_CHANGESET}" --retry-attempts=1 --retry-period=1s; then
+    exit 0
+  fi
 
-else
+  echo "--> Starting upgrade, changeset: $RIG_CHANGESET"
+
+  echo "Starting the control plane upgrade process as described at:"
+  echo "https://github.com/openebs/openebs/blob/master/k8s/upgrades/README.md"
+
+  echo "Checking the existing control plane version..."
+  FROM_VERSION=$(get_control_plane_version)
+  if [[ "$FROM_VERSION" =~ ^[0-9]+.[0-9]+.[0-9]+$ ]]; then
+    if [ "$FROM_VERSION" == "$TO_VERSION" ]; then
+      echo "The control plane is already upgraded TO_VERSION=$TO_VERSION."
+    else
+      if [ "$FROM_VERSION" == "1.4.0" ] || [ "$FROM_VERSION" == "1.5.0" ] || [ "$FROM_VERSION" == "1.7.0" ]; then
```

Comment:
If we want to be strict on expected from versions like this, I would either try and take these versions from the Makefile, or make sure the Makefile has a comment that when changing the TO_VERSION, this line of the FROM_VERSION also needs to be modified. This way someone unfamiliar with all this code would know when bumping the destination version where to go to make other related changes.